numba/core/errors.py

```diff
@@ -425,20 +426,29 @@ def deprecate_moved_module(old_module, new_module, stacklevel=3):
         msg = _moved_no_replacement
     else:
         msg = _moved_msg1.format(old_module, new_module)
-    warnings.warn(msg, category=NumbaDeprecationWarning, stacklevel=stacklevel)
+    warnings.warn(msg, category=NumbaDeprecationWarning, stacklevel=stacklevel+1)
 
 
-def deprecate_moved_module_getattr(old_module, new_module):
-    """For replacing module level __getattr__ to warn users above modules moving
-    locations
-    """
-    def wrapper(attr):
-        mod = import_module(new_module)
+class _MovedModule(ModuleType):
+    def __init__(self, old_module_locals, new_module):
+        old_module = old_module_locals['__name__']
+        super().__init__(old_module)
+
+        # copy across dunders so that package imports work too
+        for attr, value in old_module_locals.items():
+            if attr.startswith('__') and attr.endswith('__'):
+                setattr(self, attr, value)
+
+        self.__new_module = new_module
+        deprecate_moved_module(old_module, new_module, stacklevel=3)
+
+    def __getattr__(self, attr):
+        """ warn users above modules moving locations """
+        mod = import_module(self.__new_module)
         ret_attr = getattr(mod, attr)
-        msg = _moved_msg2.format(attr, old_module, new_module)
+        msg = _moved_msg2.format(attr, self.__name__, self.__new_module)
         warnings.warn(msg, category=NumbaDeprecationWarning, stacklevel=2)
         return ret_attr
-    return wrapper
 
```

Comment:
If it's useful, I could add

```suggestion
    def __repr__(self):
        return ModuleType.__repr__(self).replace('>', ', redirecting to {}>'.format(self.__new_module))
```
Didn't seem particularly worthwhile.